### DISCUSSION TITLE
High CPU usage when Secure Keyboard Entry is enabled

### DISCUSSION THREAD

**Discussion by matevzmihalic** [ghostty-org/ghostty#3729] [Feature Requests, Ideas] [answered]
<img width="920" alt="Screenshot 2024-12-28 at 15 34 16" src="https://github.com/user-attachments/assets/f09a94ce-bd0c-4eec-87ec-215532582065" />
Ghostty is using around 20% of M1 CPU when Secure Keyboard Entry is enabled (even in just shell, without htop running).
If it's disabled or Ghostty isn't in focus (lock icon isn't displayed) the CPU drops to ~1%.

**Comment by mitchellh** [comment, marked as answer]
This is caused by the SwiftUI animation (just traced it). You can disable the icon to improve this or if someone has a better approach to doing the animation I'd be happy to do it. Given theres a config to disable it (usually for people to keep it on the all the time) and otherwise its only open a short period of time, I'm going to not consider this a bug for now. 

Thanks for pointing it out though.

**Reply by brentschroeter** [reply]
I believe the animation is also responsible for rendering issues discussed in #7333 , #9590 , and #9617 . The animation can cause the main thread's `DispatchQueue` to lag, which interferes with multiple operations including the `IOSurface.setSurfaceAsync()` call that pushes frame updates to the SwiftUI surface.

Implementing the animation effect using a [radial gradient mask + blur](https://github.com/brentschroeter/ghostty/commit/7e90e26ae1a422d3e4b62fac5eb2928be3cc74b4) instead of an inner shadow drops CPU usage by ~35% (as tested on M2 MacBook Air), while maintaining essentially the same appearance. On my machines, this resolves all 3 of the bugs discussed above and generally improved terminal responsiveness while macOS Secure Input is active.

Any feedback on further improving the animation code is appreciated! In particular, I tested it in light and dark OS modes and with light and dark terminal themes, but if there are other variations to consider then I want to be sure to test those too.

Screen grabs—v1.2.3 on the left and new animation on the right:


https://github.com/user-attachments/assets/711ec7e1-9334-42db-be9d-5e059c64675d


https://github.com/user-attachments/assets/e848a332-9c79-4f8d-b801-04535d84c76c

**Reply by mitchellh** [reply]
This sounds good to me, thanks for looking into it!

**Comment by matevzmihalic** [comment]
If anyone else finds this problem, you can disable the secure input indicator and lower CPU consumption by adding `macos-secure-input-indication = false` to settings.
